apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore.IntegrationTests/Logging/LoggingTest.cs

```diff
@@ -0,0 +1,338 @@
+// limitations under the License.
+
+using Google.Cloud.Diagnostics.Common;
+using Google.Cloud.Diagnostics.Common.Tests;
+using Google.Cloud.Logging.Type;
+using Google.Cloud.Logging.V2;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.TestHost;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Xml;
+using Xunit;
+
+namespace Google.Cloud.Diagnostics.AspNetCore.IntegrationTests
+{
+    public class LoggingTest
+    {
+        /// <summary>Total time to spend sleeping when looking for log entries.</summary>
+        private readonly TimeSpan _timeout = TimeSpan.FromSeconds(10);
+
+        /// <summary>Time to sleep between checks for log entries.</summary>
+        private readonly TimeSpan _sleepInterval = TimeSpan.FromSeconds(2);
+
+        /// <summary>Project id to run the test on.</summary>
+        private readonly string _projectId;
+
+        /// <summary>Client to use to send RPCs.</summary>
+        private readonly LoggingServiceV2Client _client;
+
+        public LoggingTest()
+        {
+            _projectId = Utils.GetProjectIdFromEnvironment();
+            _client = LoggingServiceV2Client.Create();
+        }
+
+        /// <summary>
+        /// Gets log entries that contain the passed in testId in the log message.  Will poll
+        /// and wait for the entries to appear.
+        /// </summary>
+        /// <param name="startTime">The earliest log entry time that will be looked at.</param>
+        /// <param name="testId">The test id to filter log entries on.</param>
+        /// <param name="minEntries">The minimum number of logs entries that should be waited for.
+        ///     If minEntries is zero this method will wait the full timeout before checking for the
+        ///     values to ensure they have not appeared.</param>
+        private IEnumerable<LogEntry> GetEntries(DateTime startTime, string testId, int minEntries)
+        {
+            TimeSpan totalSleepTime = TimeSpan.Zero;
+            while (totalSleepTime < _timeout)
+            {
+                TimeSpan sleepTime = minEntries > 0 ? _sleepInterval : _timeout;
+                totalSleepTime += sleepTime;
+                Thread.Sleep(sleepTime);
+
+                // Convert the time to RFC3339 UTC format.
+                string time = XmlConvert.ToString(startTime, XmlDateTimeSerializationMode.Utc);
+                var request =  new ListLogEntriesRequest
+                {
+                    ResourceNames = { $"projects/{_projectId}" },
+                    Filter = $"timestamp >= \"{time}\""
+                };
+
+                var results = _client.ListLogEntries(request);
+                var entries = results.Where(p => p.TextPayload.Contains(testId));
```

Comment:
It may be good to end this expression with `.ToList()`. Both the `Count()` call below and all the `FirstOrDefault/Count/Where` calls in the callers are going to re-run possibly the entire enumerator, which is going to re-filter all elements with this predicate each time, as well as do whatever stuff `_client.ListLogEntries` is doing it its enumerable pipeline. `.ToList()` will hydrate the enumerate into a new list once and then that can be enumerated simply all other times.

Actually, looking into it briefly just now, it appears ListLogEntries might actually be doing a network call (I didn't look into enough to see if any caching is going on), so this might explain why you were hitting quota issues. You might have been requesting log entries far more than you intended.